projects/opencv/readnetfromtensorflow_fuzzer.cc

```diff
@@ -0,0 +1,73 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+ * This fuzzer is generated by UTopia project based on TEST(Test_Tensorflow, read_inception).
+ * (UTopia Project: https://github.com/Samsung/UTopia)
+ */
+#include <opencv2/dnn/dnn.hpp>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/imgproc.hpp>
+#include <fuzzer/FuzzedDataProvider.h>
+#include <fstream>
+
+using namespace cv;
+using namespace dnn;
+
+bool saveFile(std::string Path, std::string Content) {
+  std::ofstream OFS(Path);
+  if (!OFS.is_open())
+    return false;
+
+  OFS << Content;
+  return true;
+}
+
+static inline void fuzz(FuzzedDataProvider &Provider) {
+  auto Input1 = Provider.ConsumeRandomLengthString();
+  int Input2 = Provider.ConsumeIntegral<int>();
+  auto Input3 = Provider.ConsumeRandomLengthString();
+  std::string Input3Path = "input3";
+  if (!saveFile(Input3Path, Input3)) return;
+  int Input4 = Provider.ConsumeIntegralInRange<int>(0, 256);
+  int Input5 = Provider.ConsumeIntegralInRange<int>(0, 256);
+  int Input6 = Provider.ConsumeIntegralInRange<int>(0, 256);
+  auto Input7 = Provider.ConsumeRandomLengthString();
+  auto Input8 = Provider.ConsumeRandomLengthString();
+
+  Net net;
+  net = readNetFromTensorflow(Input1.c_str(), Input1.size());
+  if (net.empty())
+    return;
+  net.setPreferableBackend(Input2);
+
+  Mat sample = imread(Input3Path);
+  if (sample.empty())
+    return;
+
+  Mat input;
+  resize(sample, input, Size(Input4, Input5));
+  input -= Scalar::all(Input6);
+
+  Mat inputBlob = blobFromImage(input);
+
+  net.setInput(inputBlob, Input7);
+  Mat out = net.forward(Input8);
```

Comment:
It is unlikely to inference with random data (random size). It would fail in most cases due to size difference. To increase code coverage (and run layers) we need to pass valid input.

Also we should not duplicate `imread()` testing as we already have fuzzer for that.

If it is possible then we could check inference with "zero" input (or inputs). Need to extract dimensions from loaded `cv::Net` and prepare valid input (at least with zero values).

Reply:
Thanks for your comment! Unfortunately, I didn't think of the code for your suggestion. Would you suggest changes for your comments?

And let me explain about this fuzzer code simply.
This fuzzer was written based on opencv testcase. (https://github.com/opencv/opencv/blob/9da9e8244b75a3754ff3a5d07ffdac46ee28ab6b/modules/dnn/test/test_tf_importer.cpp#L30-L52).

Any constant values were changed to random data by fuzzers to try to find any possible bugs in this valid API function sequence. Especially, to invoke net.setInput(inputBlob, Input7), imread is called, because  ```inputBlob``` comes from the return value of that function. Finally, imread is also tested with random data in this context even if that API is tested in other fuzzer. As you said, large search space may cause inefficient fuzzing performance. However, I belileved that libfuzzer's coverage guided fuzzing finally find semi-valid inputs and may have more chances to visit unexpected program states (similar as coverage) to find memory related bugs.

Of course, only testing readNetfromTensorflow function is enough to focusing on fuzzing a specific function behavior like below.
```
static inline void fuzz(FuzzedDataProvider &Provider) {
  auto Input = Provider.ConsumeRandomLengthString();
  Net net = readNetFromTensorflow(Input.c_str(), Input.size());
}
```

Reply:
I believe we could start from `readNetFromTensorflow()` call only.

Reply:
4a6a788: Fixed, now only fuzzing readNetFromTensorflow().